File: pysm/experiments/evaluation_metrics/semantic_labeling.py
#!/usr/bin/python
# -*- coding: utf-8 -*-

import numpy as np
from typing import Dict, Tuple, List, Union, Optional

from data_structure import Graph
from semantic_modeling.karma.semantic_model import SemanticModel


def mrr(gold_sm: SemanticModel, pred_sm: SemanticModel) -> float:
    """The learned semantic types should be stored in pred_sm.attributes"""
    ranks = []
    for attr in gold_sm.attrs:
        pred_attr = pred_sm.label2attrs[attr.label]
        if not gold_sm.graph.has_node_with_id(attr.id):
            # this column is ignored, and not used in sm
            if len(pred_attr.semantic_types) == 0:
                ranks.append(1)
            else:
                ranks.append(0)
        else:
            node = gold_sm.graph.get_node_by_id(attr.id)
            assert node.n_incoming_links == 1, "Not support multi-parents"

            link = node.get_first_incoming_link()
            gold_st_domain = link.get_source_node().label.decode("utf-8")
            gold_st_type = link.label.decode("utf-8")

            for i, pred_st in enumerate(pred_attr.semantic_types):
                if pred_st.domain == gold_st_domain and pred_st.type == gold_st_type:
                    ranks.append(1.0 / (i + 1))
                    break
            else:
                ranks.append(0)

    return np.average(ranks)


def accuracy(gold_sm: SemanticModel, pred_sm: SemanticModel) -> float:
    """The learned semantic types should be stored in pred_sm.attributes"""
    ranks = []
    for attr in gold_sm.attrs:
        pred_attr = pred_sm.label2attrs[attr.label]
        if not gold_sm.graph.has_node_with_id(attr.id):
            # this column is ignored, and not used in sm
            if len(pred_attr.semantic_types) == 0:
                ranks.append(1)
            else:
                ranks.append(0)
        else:
            node = gold_sm.graph.get_node_by_id(attr.id)
            assert node.n_incoming_links == 1, "Not support multi-parents"

            link = node.get_first_incoming_link()
            gold_st_domain = link.get_source_node().label.decode("utf-8")
            gold_st_type = link.label.decode("utf-8")

            if len(pred_attr.semantic_types) == 0:
                ranks.append(0)
            else:
                pred_st = pred_attr.semantic_types[0]
                if pred_st.domain == gold_st_domain and pred_st.type == gold_st_type:
                    ranks.append(1)
                else:
                    ranks.append(0)

    return np.average(ranks)


def accuracy_version2(gold_sm: Graph, pred_sm: Graph) -> float:
    dnodes = {}
    for dnode in gold_sm.iter_data_nodes():
        dlink = dnode.get_first_incoming_link()
        dnodes[dnode.label] = (dlink.get_source_node().label, dlink.label)

    assert len(dnodes) == sum(1 for n in gold_sm.iter_data_nodes()), "Label of data nodes must be unique"
    prediction = {}
    for dnode in pred_sm.iter_data_nodes():
        dlink = dnode.get_first_incoming_link()
        if dlink is None:
            continue
            
        if (dlink.get_source_node().label, dlink.label) == dnodes[dnode.label]:
            prediction[dnode.label] = 1
        else:
            prediction[dnode.label] = 0

    for lbl in dnodes:
        if lbl not in prediction:
            prediction[lbl] = 0

    assert len(prediction) == len(dnodes)
    return sum(prediction.values()) / len(prediction)


def coverage(gold_sm: SemanticModel, pred_sm: SemanticModel) -> float:
    """The learned semantic types should be stored in pred_sm.attributes"""
    ranks = []
    for attr in gold_sm.attrs:
        pred_attr = pred_sm.label2attrs[attr.label]
        if not gold_sm.graph.has_node_with_id(attr.id):
            # this column is ignored, and not used in sm
            if len(pred_attr.semantic_types) == 0:
                ranks.append(1)
            else:
                ranks.append(0)
        else:
            node = gold_sm.graph.get_node_by_id(attr.id)
            assert node.n_incoming_links == 1, "Not support multi-parents"

            link = node.get_first_incoming_link()
            gold_st_domain = link.get_source_node().label.decode("utf-8")
            gold_st_type = link.label.decode("utf-8")

            if len(pred_attr.semantic_types) == 0:
                ranks.append(0)
            else:
                for pred_st in pred_attr.semantic_types:
                    if pred_st.domain == gold_st_domain and pred_st.type == gold_st_type:
                        ranks.append(1)
                        break
                else:
                    ranks.append(0)

    return np.average(ranks)

File: mira/src/evaluation_metrics/semantic_labeling.rs
// use std::iter::Sum;
use std::collections::HashMap;
use algorithm::prelude::Graph;
use rdb2rdf::prelude::*;

pub fn mrr(gold_sm: &SemanticModel, pred_sm: &SemanticModel) -> f64 {
    // The learned semantic types should be stored in pred_sm.attributes
    let mut ranks = vec![];
    for attr in &gold_sm.attrs {
        let pred_attr = pred_sm.get_attr_by_label(&attr.label);
        if !gold_sm.graph.has_node_with_id(attr.id) {
            // this column is ignored, and not used in sm
            if pred_attr.semantic_types.len() == 0 {
                ranks.push(1.0);
            } else {
                ranks.push(0.0)
            }
        } else {
            let node = gold_sm.graph.get_node_by_id(attr.id);
            assert_eq!(node.n_incoming_edges, 1, "Not support multi-parents");
            let edge = node.first_incoming_edge(&gold_sm.graph).unwrap();
            let gold_st_domain = &edge.get_source_node(&gold_sm.graph).label;

            let mut found_in_prediction = false;
            for (i, pred_st) in pred_attr.semantic_types.iter().enumerate() {
                if &pred_st.class_uri == gold_st_domain && pred_st.predicate == edge.label {
                    ranks.push(1.0 / (i + 1) as f64);
                    found_in_prediction = true;
                    break;
                }
            }
            if !found_in_prediction {
                ranks.push(0.0);
            }
        }
    }

    ranks.iter().sum::<f64>() / ranks.len() as f64
}

pub fn accuracy(gold_sm: &SemanticModel, pred_sm: &SemanticModel) -> f64 {
    // The learned semantic types should be stored in pred_sm.attributes
    accuracy_(&gold_sm.graph, &pred_sm.graph)
}

pub(super) fn accuracy_(gold_sm: &Graph, pred_sm: &Graph) -> f64 {
    // The learned semantic types should be stored in pred_sm.attributes
    let mut dnodes: HashMap<String, (String, String)> = Default::default();
    for dnode in gold_sm.iter_data_nodes() {
        let dlink = dnode.first_incoming_edge(&gold_sm).unwrap();
        dnodes.insert(dnode.label.clone(), (dlink.get_source_node(&gold_sm).label.clone(), dlink.label.clone()));
    }

    assert_eq!(dnodes.len(), gold_sm.iter_data_nodes().count(), "Label of data nodes must be unique");
    let mut prediction: HashMap<_, _> = Default::default();
    for dnode in pred_sm.iter_data_nodes() {
        let dlink = dnode.first_incoming_edge(&pred_sm).unwrap();
        let dstypes = &dnodes[&dnode.label];

        if &dstypes.0 == &dlink.get_source_node(&pred_sm).label && &dstypes.1 == &dlink.label {
            prediction.insert(dnode.label.clone(), 1.0);
        } else {
            prediction.insert(dnode.label.clone(), 0.0);
        }
    }

    for lbl in dnodes.keys() {
        if !prediction.contains_key(lbl) {
            prediction.insert(lbl.clone(), 0.0);
        }
    }

    assert_eq!(prediction.len(), dnodes.len());
    prediction.values().sum::<f64>() / prediction.len() as f64
}

pub fn coverage(gold_sm: &SemanticModel, pred_sm: &SemanticModel) -> f64 {
    // The learned semantic types should be stored in pred_sm.attributes
    let mut ranks = vec![];
    for attr in &gold_sm.attrs {
        let pred_attr = pred_sm.get_attr_by_label(&attr.label);
        if !gold_sm.graph.has_node_with_id(attr.id) {
            // this column is ignored, and not used in sm
            if pred_attr.semantic_types.len() == 0 {
                ranks.push(1.0);
            } else {
                ranks.push(0.0)
            }
        } else {
            let node = gold_sm.graph.get_node_by_id(attr.id);
            assert_eq!(node.n_incoming_edges, 1, "Not support multi-parents");
            let edge = node.first_incoming_edge(&gold_sm.graph).unwrap();
            let gold_st_domain = &edge.get_source_node(&gold_sm.graph).label;

            let mut found_in_prediction = false;
            for (i, pred_st) in pred_attr.semantic_types.iter().enumerate() {
                if &pred_st.class_uri == gold_st_domain && pred_st.predicate == edge.label {
                    ranks.push(1.0);
                    found_in_prediction = true;
                    break;
                }
            }
            if !found_in_prediction {
                ranks.push(0.0);
            }
        }
    }

    ranks.iter().sum::<f64>() / ranks.len() as f64
}